### PR TITLE
Chore: shipping methods tax inclusive total

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
What:
Calculate taxes from shipping methods where the tax is included in the price (includes_tax = true).

FIXES CORE-376

### CODE REVIEW

File: packages/medusa/src/strategies/tax-calculation.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/strategies/tax-calculation.ts:101]: The If/Else like this has the purpose of making it easy when cleaning up the FF

File: packages/medusa/src/strategies/__tests__/tax-calculation.js
Comment (adrien2p) [inline packages/medusa/src/strategies/__tests__/tax-calculation.js:276]: comment: you will be able to clean that part after this pr is merged https://github.com/medusajs/medusa/pull/2136

File: packages/medusa/src/strategies/tax-calculation.ts
Comment (adrien2p) [inline packages/medusa/src/strategies/tax-calculation.ts:101]: comment: same as the previous one

File: packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts
Comment (adrien2p) [inline packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts:7]: comment: same as above

**Review by adrien2p — COMMENTED** [review]
Nice work man 👍

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:211]: question: I am wondering if it really matters to know if it is a cart or an order in the statements as it does not change the logic

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:238]: todo: listing issue here on my side.

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:211]: question: previously if it was an order with a tax rate, we were calculating the taxes with the provided tax rate to keep the old system working.

Now we have lost that feature, wdyt?

File: packages/medusa/src/services/totals.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/totals.ts:316]: Not sure if I understood it. In case we don't have tax_lines, the taxes will be 0, and free shipping will have price 0 no?

Comment (adrien2p) [reply packages/medusa/src/services/totals.ts:316]: if we have a free shipping discount then it will be 0, but I am wondering why it is not taken into account for the shipping total. specially that the discount total, filter on the discounts that are not free shipping.

also, if there is not tax lines it could be due to the back compatibility of the tax rate being setup on the cartOrOrder, but it is not taken into account. now that there is the inclusive tag, the best way to know the real price of the shipping method would be to call the method `getShippingMethodTotals` that take care of the different circumstances (plus the comment on the method too). that way you can compute back the real price. it is not the more performant at the moment, but until it is refactored i think it is the best approach, a bit like i ve done it here `getLineItemTotal`

@srindom could you confirm if i am missing something just for that part please?

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:252]: suggestion: move and store that in a variable above the if and combine such as 
```ts
    const hasFreeShipping =
      !!cartOrOrder.discounts &&
      cartOrOrder.discounts.some(
        (d) => d.rule.type === DiscountRuleType.FREE_SHIPPING
      )
    if (hasFreeShipping) {
      totals.total = 0
      totals.tax_total = 0
    }
```

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:213]: @srindom can you confirm that it was a bug and it should not be original_tax_total * tax_rate but instead price * tax_rate?

Comment (srindom) [reply packages/medusa/src/services/totals.ts:213]: Yes you are right!

Comment (adrien2p) [reply packages/medusa/src/services/totals.ts:213]: Thanks, in that case it answer my comment here https://github.com/medusajs/medusa/pull/2130#discussion_r961804701 👌

File: packages/medusa/src/strategies/tax-calculation.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/strategies/tax-calculation.ts:101]: This FF is ok here, before calling calculatePriceTaxAmount where the FF will be removed.
The if/else block is to avoid `includesTax: FF && sm.includes_tax ?? false`, making it easier when we clean up the FF.

Comment (adrien2p) [reply packages/medusa/src/strategies/tax-calculation.ts:101]: I don’t think the` ?? false` is necessary in your example. It would be` !!(FF && sm.includes_tax)` wdyt? The thing is that if you store it before the loop, it would avoid having the re execute it for each loop knowing that the value does not change. 

I am fine with both but i would like your opinion.

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/strategies/tax-calculation.ts:101]: True. but most places where we have `?? false` I would also remove them. I'm fine with checking `if(undefined)` instead of forcing `if(false)`. 
But moving the FF outside the loop is ok to me.

File: packages/medusa/src/services/totals.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/totals.ts]: fixed

File: packages/medusa/src/strategies/tax-calculation.ts
Comment (adrien2p) [reply packages/medusa/src/strategies/tax-calculation.ts]: Ahah this is not what i meant. I ll write the example tomorrow or later when i get to my computer ☺️

File: packages/medusa/src/services/pricing.ts
Comment (adrien2p) [inline packages/medusa/src/services/pricing.ts:138]: comment(non blocking): my linter does not arrange it that way. is that coming from my conf?
```
      taxedPricing.calculated_price_incl_tax =
        variantPricing.calculated_price_includes_tax
          ? variantPricing.calculated_price
          : variantPricing.calculated_price + taxedPricing.calculated_tax
```

Comment (adrien2p) [reply packages/medusa/src/services/pricing.ts:138]: it the same below

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:934]: todo: I think there has been a problem during your merge. it should not be that way

Comment (adrien2p) [reply packages/medusa/src/services/totals.ts:934]: just to be sure could you check that there is no other un expected merge problem please :)

File: packages/medusa/src/services/totals.ts
Comment (adrien2p) [inline packages/medusa/src/services/totals.ts:1085]: praise: Make sense to use it for the next constraint since if there is no discount we will not loop anyway 👍

File: packages/medusa/src/strategies/__tests__/tax-calculation.js
Comment (adrien2p) [reply packages/medusa/src/strategies/__tests__/tax-calculation.js:277]: in the last version of the base branch, this part should not be there. You should get the following:
```ts
describe("TaxCalculationStrategy", () => {
  describe("calculate", () => {
    test.each(toTest)(
      "%s",
      async (title, { items, taxLines, context, expected, flags }) => {
        const featureFlagRouter = new FlagRouter(flags ?? {})
        const calcStrat = new TaxCalculationStrategy({
          featureFlagRouter
        })

        const val = await calcStrat.calculate(items, taxLines, context)
        expect(val).toEqual(expected)
      }
    )
  })
})
```

File: packages/medusa/src/strategies/tax-calculation.ts
Comment (adrien2p) [inline packages/medusa/src/strategies/tax-calculation.ts:104]: suggestion(non blocking): wdyt?
```ts
    const taxInclusiveEnabled = this.featureFlagRouter_.isFeatureEnabled(
      TaxInclusivePricingFeatureFlag.key
    )

    let taxTotal = 0
    for (const sm of shipping_methods) {
      const lineRates = taxLines.filter((tl) => tl.shipping_method_id === sm.id)
      for (const lineRate of lineRates) {
        taxTotal += calculatePriceTaxAmount({
          price: sm.price,
          taxRate: lineRate.rate / 100,
          includesTax: taxInclusiveEnabled && sm.includes_tax,
        })
      }
    }
    return taxTotal
```

File: packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts
Comment (adrien2p) [reply packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts:7]: should not be necessary with the merge

File: packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts
Comment (adrien2p) [inline packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts:34]: suggestion: rm `toFixed` and use the assertion `toBeCloseTo` instead

File: packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts
Comment (adrien2p) [inline packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts:44]: suggestion: rm `toFixed` and use the assertion `toBeCloseTo` instead

File: packages/medusa/src/services/__tests__/totals.js
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/__tests__/totals.js:5]: I use it just to mock `spyOn(CalculatePriceTaxAmount, "calculatePriceTaxAmount")`

File: packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/utils/__tests__/calculate-price-tax-amount.ts:34]: Nice call

File: packages/medusa/src/services/pricing.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/pricing.ts:138]: My prettier is doing that.
Double check ESLint and Prettier are in my checklist. At some point we have to normalize all the files in the project.

File: packages/medusa/src/services/totals.ts
Comment (srindom) [reply packages/medusa/src/services/totals.ts:321]: Just to weigh in here:

Think it is a great idea to change this method to just be:
   ```js
    const { shipping_methods } = cartOrOrder
    return shipping_methods.reduce((acc, next) => {
      const totals = this.getShippingMethodTotals(next, cartOrOrder)
      return acc + totals.subtotal
    }, 0)
   ```
   
It will then be `getShippingMethodTotals` that is responsible for doing the tax calculation stuff.

Not sure if that was what you meant @adrien2p?

Comment (adrien2p) [reply packages/medusa/src/services/totals.ts:321]: Yes this is exactly what i meant 😊 thanks man

Comment (adrien2p) [reply packages/medusa/src/services/totals.ts:321]: But surrounded by a lot of explaination on why that choice which might be less understandable in the end 🤔

Comment (srindom) [reply packages/medusa/src/services/totals.ts:321]: Haha alright :D 

@carlos-r-l-rodrigues - does this make sense to you?

File: packages/medusa/src/services/order.ts
Comment (srindom) [inline packages/medusa/src/services/order.ts:169]: todo: my linter wants this to be like below - can you check your setup?

```js
    const { select, relations, totalsToSelect } =
      this.transformQueryForTotals(config)
```

Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/services/order.ts:169]: Yeah, apparently ESlint and Prettier are behaving differently. I will check the rules, but I hope this is not blocking the PR for now.
At some point we have to apply a lint --fix to the entire project and validate it as the first step in the pipeline.

Comment (adrien2p) [reply packages/medusa/src/services/order.ts:169]: 👍

**Review by srindom — APPROVED** [review]
LGTM! 🎉

**Pull request merged**
